services/src/main/java/edu/unc/lib/dl/cdr/services/rest/modify/IndexingController.java

```diff
@@ -15,104 +15,116 @@
+    }
 
-        if (!hasPermission(id)) {
-            response.setStatus(401);
-            return;
-        }
+        private ResponseEntity<Object> indexObject(String id) {
+            Map<String, Object> result = new HashMap<>();
+            result.put("action", "reindex");
+            result.put("pid", id);
 
-        log.info("Updating " + id);
-        operationsMessageSender.sendIndexingOperation(GroupsThreadStore.getUsername(),
-                Arrays.asList(pid), IndexingActionType.ADD);
-    }
+            PID objectPid = PIDs.get(id);
+
+            log.info("Updating object " + id);
+            try {
+                indexingService.reindexObject(AgentPrincipals.createFromThread(), objectPid);
+            } catch (Exception e) {
+                result.put("error", e.getMessage());
+                if (e instanceof AuthorizationException || e instanceof AccessRestrictionException) {
+                    return new ResponseEntity<>(result, HttpStatus.FORBIDDEN);
+                } else {
+                    log.error("Failed to reindex repository due to {}",  e);
+                    return new ResponseEntity<>(result, HttpStatus.INTERNAL_SERVER_ERROR);
+                }
+            }
 
-    private boolean hasPermission(String id) {
-        // Disallow requests by users that are not at least curators for pid
-        AccessGroupSet groups = GroupsThreadStore.getGroups();
-        if (log.isDebugEnabled()) {
-            log.debug("hasPermission for groups " + groups.joinAccessGroups(";"));
+            result.put("timestamp", System.currentTimeMillis());
+            return new ResponseEntity<>(result, HttpStatus.OK);
         }
-        return groups.contains(AccessGroupConstants.ADMIN_GROUP);
-    }
 
-    public void setOperationsMessageSender(OperationsMessageSender operationsMessageSender) {
-        this.operationsMessageSender = operationsMessageSender;
-    }
+        private ResponseEntity<Object> indexObjectAndChildren(String id, Boolean inplace) {
+            Map<String, Object> result = new HashMap<>();
+            result.put("action", "reindex");
+            result.put("pid", id);
+
+            PID objectPid = PIDs.get(id);
+
+            if (inplace == null || inplace) {
+                log.info("Reindexing " + id + ", inplace reindex mode");
+            } else {
+                log.info("Reindexing " + id + ", clean reindex mode");
+            }
+            try {
+                indexingService.reindexObjectAndChildren(AgentPrincipals.createFromThread(), objectPid, inplace);
+            } catch (Exception e) {
+                result.put("error", e.getMessage());
+                Throwable t = e.getCause();
+                if (t instanceof AuthorizationException || t instanceof AccessRestrictionException) {
+                    return new ResponseEntity<>(result, HttpStatus.FORBIDDEN);
+                } else {
+                    log.error("Failed to reindex repository due to {}",  e);
```

Comment:
"Failed to reindex {} object and its children"

Reply:
done